libc/string_mem.c

```diff
@@ -0,0 +1,396 @@
+	/* Move somewhere between */
+	ret = memmove(ptr, mem + OFFSET, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(ptr, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ptr, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, mem + OFFSET, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem, OFFSET);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem + OFFSET + TEST_STR_SIZE, move_offset - (OFFSET + TEST_STR_SIZE));
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem + move_offset + TEST_STR_SIZE, MEM_SIZE - (move_offset + TEST_STR_SIZE));
+}
+
+
+TEST(string_memmove, no_overlap_adjacent)
+{
+	uint8_t *ptr = mem + OFFSET + TEST_STR_SIZE;
+	/* Move so that memory blocks are next to each other*/
+	ret = memmove(ptr, mem + OFFSET, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(ptr, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ptr, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, mem + OFFSET, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem, OFFSET);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem + OFFSET + (TEST_STR_SIZE * 2), MEM_SIZE - (OFFSET + (TEST_STR_SIZE * 2)));
+}
+
+
+TEST(string_memmove, whole_overlap)
+{
+	uint8_t *ptr = mem + OFFSET;
+	/* Move at the same address as block is */
+	ret = memmove(ptr, mem + OFFSET, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(ptr, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ptr, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem, OFFSET);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem + OFFSET + TEST_STR_SIZE, MEM_SIZE - (OFFSET + TEST_STR_SIZE));
+}
+
+
+TEST(string_memmove, right_overlap)
+{
+	uint8_t *ptr = mem + (OFFSET + OVERLAP_LEN);
+	/* Move data so that data blocks overlap */
+	ret = memmove(ptr, mem + OFFSET, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(ptr, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ptr, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, mem + OFFSET, TEST_STR_SIZE - OVERLAP_LEN);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem, OFFSET);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem + OFFSET + TEST_STR_SIZE + OVERLAP_LEN, MEM_SIZE - (OFFSET + TEST_STR_SIZE + OVERLAP_LEN));
+}
+
+
+TEST(string_memmove, left_overlap)
+{
+	uint8_t *ptr = mem + (OFFSET - OVERLAP_LEN);
+	/* Move data so that data blocks overlap */
+	ret = memmove(ptr, mem + OFFSET, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(ptr, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ptr, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR + OVERLAP_LEN, ptr + TEST_STR_SIZE, OVERLAP_LEN);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem, OFFSET - OVERLAP_LEN);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem + OFFSET + TEST_STR_SIZE, MEM_SIZE - (OFFSET + TEST_STR_SIZE));
+}
+
+
+TEST(string_memmove, move_nothing)
+{
+	uint8_t *ptr = mem + OFFSET;
+	/* Check if any data is moved */
+	ret = memmove(ptr, TEST_STR, 0);
+	TEST_ASSERT_EQUAL_PTR(ptr, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ptr, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem, OFFSET);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem + OFFSET + TEST_STR_SIZE, MEM_SIZE - (OFFSET + TEST_STR_SIZE));
+}
+
+
+TEST(string_memmove, move_byte)
+{
+#if __BYTE_ORDER == __LITTLE_ENDIAN
```

Comment:
nah, no need for a test like that, If You reaaaly want to test if one-byte memmove works that way, use `uint8_t mem[4]`

SIde note: use EQUAL_UINT32 for unsigned type